### ISSUE TITLE
Ghost Doctor improvements

### ISSUE THREAD

**Issue by acburdine** [TryGhost/Ghost-CLI#47]
When speccing out the CLI and its requirements, one of the bigger ones was ensuring that the CLI would have the least chance of failure when installing or updating Ghost. This will hopefully reduce the number of support requests that have to be addressed. 😄 

First, we are establishing the idea of an "official" Ghost stack (see TryGhost/Ghost#7414). However, because there are still oddities that can happen, even with an official environment, more things are needed to ensure the greatest chance of install/update success.

Thus, the `ghost doctor` command exists. This command is designed to run several checks on the system before anything major happens, and the hope is to catch any potential errors _before_ they happen so a more clear error message, as well as a path to fixing/avoiding the errors, can be output.

The `ghost doctor` command also has a concept of "categories", which determine when the specific check is to be run. Essentially, it has one set of checks to be run before installing Ghost for the first time, one set to run before starting ghost, and another set to be run before attempting to update Ghost.

There are already several checks in `ghost doctor` that were copied over from the Ghosts' startup check file. However, there are more that need to be created. They are sorted by category here for convenience
## Install Checks (to be run before `ghost install` does anything)
- ~~check system node version~~ (already completed)
- ~~check file permissions~~ (already completed)
## Startup Checks (to be run before `ghost start`)
- [x] Validate Config (see #26)

## Future improvements!
Although NPM generally does a pretty good job of installing the dependencies that Ghost requires, there are on occasion various glitches that result in different dependencies not being installed correctly. This check would recursively iterate through each dependency and ensure all of the dependencies are installed correctly.

Ghost generally takes ~ 100mb of memory to run well, and just as a sanity check it would probably be a good idea to ensure there is enough free memory on the system to run Ghost. (Node's `os` module has a `freemem()` method that should work for this purpose
## Update Checks (to be run before `ghost update`)
~~- [ ] System has enough memory to run NPM (requires additional research)~~

~~NPM itself can eat up a lot of system memory, and what we _don't_ want to happen is for npm to eat up too much memory and end up crashing an already-running Ghost blog. Similar to the above memory check, the `os.freemem()` method can be used, but additional research is required to see just how much memory npm uses.~~

With the switch to Yarn in #138 - the install process uses significantly less memory, making the necessity of this probably not as great.

## Tasks

See also [comment notes](https://github.com/TryGhost/Ghost-CLI/issues/47#issuecomment-357439800).

### prio
- [x] test current state of the permission checks PR #613 
- [x] detect installations with root -> this is wrong (only for install checks, never for update) PR #604 
- [x] detect if somebody is running `sudo ghost start` PR #604 
  - doing this in the Command base class, detect process.uid = 0
- [x] content folder permission PR #610 
  - run on ghost start
  - if content folder is _not_ owned by ghost, skip
  - all subfolders of the content folder need to be owned by ghost
  - all file & folder permissions need to be correct - see #294 for which permissions should be set
  - also, check error messaging when content folder is _not_ owned by ghost

### later
- [ ] repair permissions, #317 and #294 
- [x] memory detection before `ghost update` (relates to https://github.com/TryGhost/Ghost-CLI/issues/192)
- [ ] [optional] Check installed dependencies -> needs spec cc @acburdine 
- [x] make it possible to run doctor checks for extensions
- [ ] check that cronjob was configured correctly #554 
- [x] auto-run ghost doctor before specific commands (be able to disable it) -> do after we're able to group tasks for different commands
- [ ] be able to skip the doctor (see [here](https://github.com/TryGhost/Ghost-CLI/issues/47#issuecomment-357439800) and [here](https://github.com/TryGhost/Ghost-CLI/issues/654#issuecomment-366851777))
- [x] similar to root user prevention, detect when a user installed Ghost as `ghost` and/or is logged in as such. See [here](https://github.com/TryGhost/Ghost-CLI/issues/47#issuecomment-366219514) -> PR #659 & 657
- [x] check `.ghost-cli` permissions and ownership. See [here](https://github.com/TryGhost/Ghost-CLI/issues/47#issuecomment-366260664)
- [x] check `/versions` directory permissions and ownership. When updating ghost and removing older versions, it might fail
- [x] detect when logged in user is not the owner of the ghost installation folder and/or the logged in user is not part of the group of the owner user. -> PR #659 
- [ ] detect wrong symlink permissions (https://github.com/TryGhost/Ghost-CLI/pull/737)
- [ ] https://forum.ghost.org/t/trying-to-upgrade-from-1-8-6/2318/14

**Comment by PaszaVonPomiot** [comment]
@acburdine could you maybe add the step for ghost doctor to verify current ghost instance (or better all instances) against `~/.ghost/config` file? Also is it possible to automatically update `~/.ghost/config` file with correct Ghost instance data?

I often get into following situation:

```
root@ghost:/var/www/ghost# ghost start
 ✔ Validating config
✔ Starting Ghost
You can access your blog at http://192.168.1.75:2368/
root@ghost:/var/www/ghost# ghost ls
┌──────┬──────────┬────────┬─────┬──────┬─────────────────┐
│ Name │ Location │ Status │ Url │ Port │ Process Manager │
└──────┴──────────┴────────┴─────┴──────┴─────────────────┘
root@ghost:/var/www/ghost#
```

**Comment by kirrg001** [comment]
1. extend or add ability to fix file/folder permissions for the user.
2. run ghost doctor automatically before certain commands (start/update/restart - maybe more) - prevention of running into errors -> if ghost doctor fails, it should error and suggests running `ghost doctor --fix`
3. be able to skip doctor execution (e.g. `--skip doctor`)
3. ghost doctor just detects whats wrong, doesn't do anything by default

**Comment by AileenCGN** [comment]
Additional check to add from https://github.com/TryGhost/Ghost-CLI/issues/575#issuecomment-359157945:

- check installation folder (needs to be `/var/www`)

**Comment by AileenCGN** [comment]
@kirrg001 @acburdine 

> detect installations with root -> this is wrong (only for install checks, never for update)

is this what we want to achieve for this ^ task:

![image](https://user-images.githubusercontent.com/8037602/35396194-8399a434-01ec-11e8-9a59-d3dce60cc243.png)

I will skip this task when it's installed locally.

**Comment by AileenCGN** [comment]
For the last task of the prioritised improvements, I will focus on this:

> test current state of the permission checks

Todos:

### Ghost install
- [ ] install ghost outside of `/var/www/` e. g. the home directory
- [ ] change permissions for `others` in directories above the ghost installation and 
    - [ ] check existing error messages, and maybe improve them
    - [ ] validate, that existing permission checks work on install command
    - [ ] see, if there's improvement we can do

### Ghost start/update/run
- [ ] change permissions inside of the ghost installation folder, to test #294 and
    - [ ] check existing error messages
    - [ ] find improvement
- [ ] change permissions for `others` in directories above the ghost installation and check behaviour (#294)
    - [ ] what happens with config?

### Note for later (after ghost doctor refactor)
- [ ] run `folderPermissions` check and `checkDirectoryAndAbove` check on the important commands like `start`, `update` 
- [ ] how can we make sure the files and directories have the correct permissions on install (#294)?

**Comment by kirrg001** [comment]
@acburdine We just had a case in slack. The user was logged in with `ghost` and all folders and files were owned by `ghost`. `ghost doctor` was green. 

As far as i remember is the `ghost` user not part of the sudoers group?

**Comment by vikaspotluri123** [comment]
I think we might need to add a check for `.ghost-cli` writeability (or possibly ownership) for the commands that need it
